processing/src/main/java/org/apache/druid/frame/key/ByteRowKeyComparator.java

```diff
@@ -68,83 +90,86 @@ public static int computeFirstFieldPosition(final int fieldCount)
+    int currentRunStartPosition2 = firstFieldPosition;
 
-    for (final KeyColumn column : keyColumns) {
-      if (column.order() == KeyOrder.NONE) {
-        throw new IAE("Key must be sortable");
-      }
+    // Number of fields compared till now, which is equivalent to the index of the field to compare next
+    int fieldsComparedTillNow = 0;
 
-      if (column.order() != order) {
-        ascDescRunLengths.add(runLength);
-        runLength = 0;
+    for (RowKeyComparisonRunLengths.RunLengthEntry runLengthEntry : rowKeyComparisonRunLengths.getRunLengthEntries()) {
 
-        // Invert "order".
-        order = order == KeyOrder.ASCENDING ? KeyOrder.DESCENDING : KeyOrder.ASCENDING;
+      if (runLengthEntry.getRunLength() <= 0) {
+        // Defensive check
+        continue;
       }
 
-      runLength++;
-    }
-
-    if (runLength > 0) {
-      ascDescRunLengths.add(runLength);
-    }
-
-    return ascDescRunLengths.toIntArray();
-  }
-
-  @Override
-  @SuppressWarnings("SubtractionInCompareTo")
-  public int compare(final byte[] keyArray1, final byte[] keyArray2)
-  {
-    // Similar logic to FrameComparaisonWidgetImpl, but implementation is different enough that we need our own.
-    // Major difference is Frame v. Frame instead of byte[] v. byte[].
-
-    int comparableBytesStartPosition1 = firstFieldPosition;
-    int comparableBytesStartPosition2 = firstFieldPosition;
+      if (!runLengthEntry.isByteComparable()) {
+        // Only complex types are not byte comparable. Nested arrays aren't supported in MSQ
+        assert runLengthEntry.getRunLength() == 1;
+        // 'fieldsComparedTillNow' is the index of the current keyColumn in the keyColumns list. Sanity check that its
+        // a known complex type
+        ColumnType columnType = rowSignature.getColumnType(keyColumns.get(fieldsComparedTillNow).columnName())
+                                            .orElseThrow(() -> DruidException.defensive("Expecting a complex column with known type"));
+        String complexTypeName = Preconditions.checkNotNull(
+            columnType.getComplexTypeName(),
+            "complexType must be present for comparison"
+        );
 
-    boolean ascending = true;
-    int field = 0;
+        ComplexMetricSerde serde = Preconditions.checkNotNull(
+            ComplexMetrics.getSerdeForType(complexTypeName),
+            "serde for type [%s] not present",
+            complexTypeName
+        );
 
-    for (int numFields : ascDescRunLengths) {
-      if (numFields > 0) {
-        final int nextField = field + numFields;
-        final int comparableBytesEndPosition1 = RowKeyReader.fieldEndPosition(keyArray1, nextField - 1);
-        final int comparableBytesEndPosition2 = RowKeyReader.fieldEndPosition(keyArray2, nextField - 1);
+        // Index of the next field that will get considered. Excludes the current field that we are comparing right now
+        final int nextField = fieldsComparedTillNow + 1;
```

Comment:
runLenghtEntry.getRunLen() should be used here. 
I think we can take it out of the if condition.